src/webgpu/shader/validation/decl/var_access_mode.spec.ts

```diff
@@ -0,0 +1,226 @@
+export const description = `
+7.3 var Declarations
+
+The access mode always has a default value, and except for variables in the
+storage address space, must not be specified in the WGSL source. See §13.3 Address Spaces.
+`;
+
+// TODO(4128): Validate the similar rules on pointer types.
+
+import { makeTestGroup } from '../../../../common/framework/test_group.js';
+import { keysOf } from '../../../../common/util/data_tables.js';
+import {
+  AccessMode,
+  AddressSpace,
+  AddressSpaceInfo,
+  kAccessModeInfo,
+  kAddressSpaceInfo,
+} from '../../types.js';
+import { ShaderValidationTest } from '../shader_validation_test.js';
+
+import { declareEntryPoint, ShaderStage } from './util.js';
+
+// Test address spaces that we can hold an i32 variable.
+const kOrdinaryAddressSpaces = keysOf(kAddressSpaceInfo).filter(as => as !== 'handle');
+
+export const g = makeTestGroup(ShaderValidationTest);
+
+/**
+ * @returns a WGSL declaration for an 'x' variable with the given parameters.
+ */
+function declareVarX(
+  addressSpace: AddressSpace | undefined,
+  accessMode: AccessMode | undefined
+): string {
+  const bindingPart =
+    addressSpace && kAddressSpaceInfo[addressSpace].binding ? '@group(0) @binding(0) ' : '';
+  const spacePart = addressSpace ? (addressSpace as string) : '';
+  const modePart = accessMode ? (accessMode as string) : '';
+  const comma = spacePart && modePart ? ',' : '';
+  const templatePart = spacePart || modePart ? `<${spacePart}${comma}${modePart}>` : '';
+
+  return `${bindingPart}var${templatePart} x: i32;`;
+}
+
+/**
+ * @returns a WGSL program for the test parameterization.
+ */
+function getShader(
+  p: {
+    addressSpace: AddressSpace; // What address space for the variable.
+    explicitSpace: boolean; // Should the address space be explicitly spelled?
+    accessMode?: AccessMode; // What access mode to use.
+    explicitMode: boolean; // Should the access mode be explicitly spelled?
+    stage: ShaderStage; // What shader stage to use.
+  },
+  additionalBody?: string
+): string {
+  const as_info = kAddressSpaceInfo[p.addressSpace];
+  const decl = declareVarX(
+    p.explicitSpace ? p.addressSpace : undefined,
+    p.explicitMode ? p.accessMode : undefined
+  );
+  const beforeShader = as_info.scope === 'module' ? decl : '';
+  const insideShader = as_info.scope === 'function' ? decl : '';
+  const body = `${insideShader}\n${additionalBody ? additionalBody : ''}`;
+  const entryPoint = declareEntryPoint({ stage: p.stage, body });
+  const prog = `${beforeShader}\n${entryPoint}`;
+  return prog;
```

Comment:
I think this is the same thing, but easier to read:

```suggestion

  additionalBody = additionalBody ?? '';

  switch (as_info.scope) {
    case 'module':
      return decl + '\n' + declareEntryPoint({ stage: p.stage, body: additionalBody });
    case 'function':
      return declareEntryPoint({ stage: p.stage, body: decl + '\n' + additionalBody });
  }
```